File: 07a.py
#! /usr/bin/python3
import sys
import re

cwd = ['']
sizes = dict()
with open(sys.argv[1], 'r') as fh:
    for line in fh:
        if re.match('\$ cd \.\.$', line):
            if len(cwd) > 1:
                cwd.pop()
        elif re.match('\$ cd /$', line):
            cwd = ['']
        elif re.match('\$ cd .+$', line):
            dir = line.split()[2]
            cwd += [dir]
        elif not re.match('dir |\$ ls', line):
            size_s, name = line.split(maxsplit=2)
            size = int(size_s)
            for i in range(1, 1 + len(cwd)):
                path = '/'.join(cwd[0:i])
                if path in sizes:
                    sizes[path] += size
                else:
                    sizes[path] = size

sum = 0
for size in sizes.values():
    if size <= 100000:
        sum += size
print(sum)

File: 07a.pl
#! /usr/bin/perl
use warnings;
use strict;
use feature qw{ say };

use ARGV::OrDATA;

my @cwd = ("");
my %dirsize;

while (<>) {
    chomp;
    if (m{^\$ cd \.\.$}) {
        pop @cwd if @cwd > 1;
    } elsif (m{^\$ cd /$}) {
        @cwd = ("");
    } elsif (m{^\$ cd (.+)}) {
        push @cwd, $1;
    } elsif (/^([0-9]+) (.+)/) {
        my ($size, $name) = ($1, $2);
        $dirsize{ join '/', @cwd[0 .. $_] } += $size for 0 .. $#cwd;
    }
}

my $sum = 0;
for my $size (values %dirsize) {
    $sum += $size if $size <= 100_000;
}
say $sum;

__DATA__
$ cd /
$ ls
dir a
14848514 b.txt
8504156 c.dat
dir d
$ cd a
$ ls
dir e
29116 f
2557 g
62596 h.lst
$ cd e
$ ls
584 i
$ cd ..
$ cd ..
$ cd d
$ ls
4060174 j
8033020 d.log
5626152 d.ext
7214296 k
